cpp/src/arrow/device.h

```diff
@@ -29,6 +29,25 @@
 
 namespace arrow {
 
+/// \brief EXPERIMENTAL: Device type enum which matches up with C Data Device types
+enum class DeviceType : char {
+  UNKNOWN = 0,
```

Comment:
Hmm... we might want to use the `kSomething` convention for these enum values, under the assumption that some software might use those uppercase names for C macros?

Reply:
Also, what does `UNKNOWN` mean here? It doesn't appear in `abi.h`.

Reply:
sure, I'll add the `k` prefix. And `UNKNOWN` was my way of handling when we don't have a device type set, i think i can update this to use your suggestion of `std::optional` to handle this instead.